File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.cpp
class SmallestInfiniteSet {
    priority_queue<int, vector<int>, greater<int>> minHeap;
    unordered_set<int> mySet;
public:
    SmallestInfiniteSet() {
        for(int i = 1; i <= 1000; ++i) {
            mySet.insert(i);
            minHeap.push(i);
        }
    }
    
    int popSmallest() {
        int x = minHeap.top();
        
        mySet.erase(x);
        minHeap.pop();
        
        return x;
    }
    
    void addBack(int num) {
        if(mySet.find(num) == mySet.end()) {
            mySet.insert(num);
            minHeap.push(num);
        }
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:

    def __init__(self):
        self.minHeap = []
        self.mySet = set()
        
        for i in range(1, 1001):
            heappush(self.minHeap, i)
            self.mySet.add(i)
        
    def popSmallest(self) -> int:
        x = heappop(self.minHeap)
        self.mySet.remove(x)
        return x
        
    def addBack(self, num: int) -> None:
        if num not in self.mySet:
            self.mySet.add(num)
            heappush(self.minHeap, num)
        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)